### PR TITLE
Replacing Stylelint with Eslint.

### PR DESCRIPTION
**Pull request by przemyslaw-zan**
### 🚀 Summary

Replacing Stylelint rules with now native support from Eslint.

---

### 📌 Related issues

* Closes https://github.com/ckeditor/ckeditor5-internal/issues/4267.

---

### 💡 Additional information

*Optional: Notes on decisions, edge cases, or anything helpful for reviewers.*

### CODE REVIEW

**Comment by filipsobol** [comment]
**Issues**
* **High**: `css-indent` does not actually enforce tab indentation; it only rejects lines starting with spaces. Unindented declarations like `.foo {\ncolor: hsl(...);\n}` pass today. This is weaker than the replaced Stylelint `@stylistic/indentation: "tab"` behavior. See `lib/rules/css-indent.js:46-53`; tests in `tests/rules/css-indent.js:66-90` only cover space-indented invalid cases, not missing indentation.
* **High**: `ck-content-variable-name` validates the whole declaration value instead of each `var(...)`. Any allowed or ignored variable suppresses all invalid variables in the same value, e.g., `.ck-content { width: var(--ck-content-good), var(--bad); }` passes. See `lib/rules/ck-content-variable-name.js:66-75`; tests do not cover mixed valid/invalid variables.
* **Medium**: `ck-content-variable-name` matches selectors via `selectorText.includes('.ck-content')`, so `.ck-contentish` and `.ck-content-extra` are treated as `.ck-content`. See `lib/rules/ck-content-variable-name.js:34-38`. The CSS AST already exposes `ClassSelector` nodes, so this can be made exact.
* **Medium**: `no-disallowed-color-formats` uses raw text scanning, which creates both false positives and false negatives. Examples: `color-mix(in srgb, ...)`, `var(--ck-color-srgb)`, `url(#gradient)`, `grid-template-areas: "red red"`, `font-family: Red Hat Text`, and `content: "blue"` are reported, while `RGB(0,0,0)` is missed. See `lib/rules/no-disallowed-color-formats.js:56-70`; tests in `tests/rules/no-disallowed-color-formats.js` do not cover these cases.
* **Medium**: `license-header` CSS mode identifies a license candidate by any block comment containing `copyright`, not by `@license` or by the configured header shape. A non-header copyright comment later in a CSS file can be treated as the header and replaced instead of reporting a missing top header. See `lib/rules/license-header.js:153-157`; CSS tests do not cover non-header comments or unexpected content before a CSS header.
* **Low**: CSS rules that report custom locations use `column: 0`, but `@eslint/css` locations are 1-based. This currently yields ESLint messages with `column: 0`, e.g. from `css-indent`. See `lib/rules/css-indent.js:56-59` and `lib/rules/license-header.js:162-163`.

**Simplification**
* The new CSS rules can avoid most string parsing by using the `@eslint/css` AST:
  * Use `Hash` nodes for hex colors.
  * Use `Function` nodes with `node.name.toLowerCase() === 'rgb' || 'rgba'`.
  * Use `Function` nodes named `var` and inspect the first `Identifier` child for `ck-content-variable-name`.
  * Use `ClassSelector` nodes with `name === 'ck-content'` for selector matching.
* To remove `NAMED_COLORS`, add a direct dependency on `@eslint/css-tree` and use its lexer/type data for `named-color`, or otherwise centralize/generated color data instead of maintaining the list manually. `@eslint/css` itself ships matching named color data internally, but it is not exported as a public subpath, so importing it directly would be fragile.

**Tests**
* `pnpm --filter eslint-plugin-ckeditor5-rules test` passes.
* The new tests cover happy paths and basic failures, but they miss the edge cases above, especially unindented CSS, mixed valid/invalid variables, exact selector matching, non-color value false positives, uppercase RGB functions, and CSS license-header non-header comments.

**Review by filipsobol — CHANGES_REQUESTED** [review]
See comment above.

File: packages/eslint-plugin-ckeditor5-rules/lib/rules/license-header.js [hunk 1]
```diff
@@ -139,3 +143,85 @@ function licenseHeaderRule( context, headerLines ) {
...
+		const text = sourceCode.getText( comment );
+
+		return text.startsWith( '/*' ) && text.toLowerCase().includes( '@license' );
+	} );
+
+	if ( !licenseComment ) {
+		context.report( {
+			loc: {
+				start: { line: 1, column: 1 },
+				end: { line: 1, column: 1 }
```
[the remark this answers is not in the document]
Comment (przemyslaw-zan) [reply packages/eslint-plugin-ckeditor5-rules/lib/rules/license-header.js:169]: Not a bug. ESLint's "columns are 0-indexed" applies to JS rules — ESLint's JS layer adds `+1` to columns on output. `@eslint/css` reports columns as-is, so the value in `loc.column` is what shows up in the message.

Verified with a probe rule that emits two reports on the same line, one with `column: 0` and one with `column: 1`, against a real CSS fixture via the project's pinned `@eslint/css` + ESLint:

```
2:0  error  probe column:0
2:1  error  probe column:1
```

`column: 0` produces the nonsensical `2:0`; `column: 1` correctly points at the start of line 2. That's exactly the off-by-one Filip flagged in the original review, which is why these were changed from `0` to `1`.

File: packages/eslint-plugin-ckeditor5-rules/lib/rules/css-indent.js [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+					const expectedLeading = '\t'.repeat( expectedDepth );
+
+					if ( leading === expectedLeading ) {
+						continue;
+					}
+
+					context.report( {
+						loc: {
+							start: { line: lineNumber, column: 1 },
+							end: { line: lineNumber, column: leading.length + 1 }
```
[the remark this answers is not in the document]
Comment (przemyslaw-zan) [reply packages/eslint-plugin-ckeditor5-rules/lib/rules/css-indent.js:182]: https://github.com/ckeditor/ckeditor5-linters-config/pull/115#discussion_r3320282981

File: packages/eslint-plugin-ckeditor5-rules/lib/rules/ck-content-variable-name.js
Comment (psmyrek) [inline packages/eslint-plugin-ckeditor5-rules/lib/rules/ck-content-variable-name.js:1]: This new ESLint rule changes the default behavior of the old Stylelint `ck-content-variable-name` rule. The old Stylelint rule had built-in exceptions for variables like `-suggestion-` and `-comment-`. The new ESLint rule only allows them when `ignoredVariableSubstrings` is passed in options, but the shared config enables the rule without any options.

So code like this used to pass and now fails:

```css
.ck-content {
    color: var(--ck-suggestion-marker);
}
```

If this stricter behavior is intentional, I think it should be mentioned as a breaking change. If not, the shared config should probably pass the old ignore list so the migration keeps the previous behavior.

Comment (przemyslaw-zan) [reply packages/eslint-plugin-ckeditor5-rules/lib/rules/ck-content-variable-name.js:1]: Added changelog: https://github.com/ckeditor/ckeditor5-linters-config/pull/115/changes/160a5ec3881f2a909611db075e3c2d2ddae8e96e

File: packages/eslint-plugin-ckeditor5-rules/package.json [hunk 3]
```diff
@@ -32,13 +32,15 @@
...
     "@es-joy/jsdoccomment": "^0.50.2",
+    "@eslint/css-tree": "^4.0.4",
     "enhanced-resolve": "^5.15.0",
     "resolve.exports": "^2.0.2",
     "upath": "^2.0.1",
     "validate-npm-package-name": "^6.0.0",
     "yaml": "^2.8.0"
   },
   "devDependencies": {
+    "@eslint/css": "^1.2.0",
```
Comment (filipsobol) [inline packages/eslint-plugin-ckeditor5-rules/package.json]: `v1.3.0` is already available

**Comment by filipsobol** [comment]
* `no-disallowed-color-formats.js`: the rule does not currently look inside custom property values. Because `@eslint/css` exposes these values as `Raw`, colors used there are silently skipped. This means theme variables can still be defined with formats that this rule is supposed to forbid:

  ```css
  :root {
	  --ck-color-hex: #fff;
	  --ck-color-rgb: rgb(0, 0, 0);
	  --ck-color-name: red;
  }
  ```

* `css-indent.js`: indentation is still not enforced inside multi-line custom property values. Those values are skipped as `Raw`, so incorrectly indented continuation lines pass unnoticed:

  ```css
  :root {
	  --ck-gradient: linear-gradient(
      hsl(0, 0%, 0%),
   hsl(0, 0%, 100%)
	  );
  }
  ```

* From my testing, the CSS license header check works for the commercial packages, but it doesn't seem to report anything for CSS files in core OSS packages.

**Review by filipsobol — CHANGES_REQUESTED** [review]

**Comment by przemyslaw-zan** [comment]
> From my testing, the CSS license header check works for the commercial packages, but it doesn't seem to report anything for CSS files in core OSS packages.

Works for me, could you provide reproduction steps?

```
14:34:30  ~/.../external/ckeditor5   ci/4267 ⚑ ●  pnpm run lint

> ckeditor5-root@48.1.1 lint /home/zanju/repositories/ckeditor5-commercial/external/ckeditor5
> eslint --format ./scripts/eslint-formatter.cjs --concurrency=4


/home/zanju/repositories/ckeditor5-commercial/external/ckeditor5/packages/ckeditor5-core/theme/core.css
  1:1  error  The license header is missing  ckeditor5-rules/license-header

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.

CKEditor 5 custom ESLint rules are described in the "Code style" guide in the documentation.
https://ckeditor.com/docs/ckeditor5/latest/framework/contributing/code-style.html#ckeditor-5-custom-eslint-rules

 ELIFECYCLE  Command failed with exit code 1.

```

**Comment by filipsobol** [comment]
My bad, it works as expected. I was running `pnpm run lint` from wrong directory.

File: pnpm-workspace.yaml [hunk 4]
```diff
@@ -11,6 +11,7 @@ minimumReleaseAgeExclude:
   - '@ckeditor/*'
   - '@cksource/*'
   - '*ckeditor5*'
+  - '@eslint/css' # Remove after 2026-06-01
```
Comment (filipsobol) [inline pnpm-workspace.yaml]: Can be removed.

**Review by filipsobol — CHANGES_REQUESTED** [review]
* `ck-content-variable-name.js`: the rule still does not look inside custom property values. Because `@eslint/css` exposes these values as `Raw`, invalid variables used there are silently skipped:

  ```css
  :root,
  .ck-content {
	  --ck-content-foo: var(--bad);
  }
  ```

* `no-disallowed-color-formats.js`: the rule does not currently look inside `var()` fallback values. `css-tree` exposes the fallback as `Raw`, so disallowed colors can still be used there:

  ```css
  .foo {
	  color: var(--ck-color-text, #fff);
	  background: var(--ck-color-background, rgb(0, 0, 0));
	  border-color: var(--ck-color-border, red);
  }
  ```

* `no-disallowed-color-formats.js`: custom property values are scanned without knowing whether the custom property is supposed to contain a color. This creates false positives for non-color custom properties (**this is an edge case, which we may live with if the fix is tricky or increases rule complexity a lot**):

  ```css
  :root {
	  --ck-font-family: Red Hat Text;
	  --ck-animation-name: red;
  }
  ```

* `css-indent.js`: indentation is still not enforced for the closing line of multi-line declaration values. The `);` line below passes even though it should be indented with one tab:

  ```css
  .foo {
	  background: linear-gradient(
		  hsl(0, 0%, 0%)
  );
  }
  ```

* `.changelog/20260526120000_ci_4484.md`: the changelog entry says custom-property values are not checked, but the rule now checks indentation inside custom-property parentheses. The entry should be updated.

**Comment by przemyslaw-zan** [comment]
**In this pass, a fixer for the indent rule was added.**

---

> no-disallowed-color-formats.js: custom property values are scanned without knowing whether the custom property is supposed to contain a color. This creates false positives for non-color custom properties (this is an edge case, which we may live with if the fix is tricky or increases rule complexity a lot):

Custom properties carry no declared type in CSS, so when a value is a `Raw` token there's no reliable signal telling us whether a bare keyword like `red` is a color or just an identifier (a font name in `--ck-font-family: Red Hat Text`, an animation name in `--ck-animation-name: red`), so any fix is therefore a tradeoff - we'd either drop legitimate named-color enforcement on custom-property definitions (e.g. stop flagging `--ck-color-bg: red`) or lean on a brittle name-based heuristic that still guesses wrong both ways. Since hex and `rgb()` are unambiguous and never affected, and we think this as an acceptable edge case to live with, we should leave it as it is.

File: packages/eslint-plugin-ckeditor5-rules/lib/rules/no-disallowed-color-formats.js [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+ * Line 1 of the sub-parsed value starts mid-file at `anchor.column`; later lines
+ * start at file column 1, so only line 1 needs the column anchor shift.
+ */
+function translatePosition( pos, anchor ) {
+	if ( pos.line === 1 ) {
+		return { line: anchor.line, column: anchor.column + pos.column - 1 };
+	}
+
+	return { line: anchor.line + pos.line - 1, column: pos.column };
+}
```
[the remark this answers is not in the document]
Comment (przemyslaw-zan) [reply packages/eslint-plugin-ckeditor5-rules/lib/rules/no-disallowed-color-formats.js:167]: This repository does not yet have a convention for utils shared between rules. This would also be a very small gain in the name of the D.R.Y. rule. Two occurrences is something we can live with for now.

**Review by filipsobol — APPROVED** [review]
